### PR TITLE
Report on System Integrity Protection

### PR DESCRIPTION
**Pull request by sharvilshah**
This is still a work in progress, but I wanted to open it up for review, especially regarding the nomenclature and the table schema.

I would also like to report on SIP exceptions defined, but that probably is another table `sip_exceptions` perhaps?

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

File: osquery/tables/system/darwin/sip_config.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+#include <osquery/logger.h>
+
+namespace osquery {
+namespace tables {
+
+// rootless configuration flags
+// https://opensource.apple.com/source/xnu/xnu-3248.20.55/bsd/sys/csr.h
+const std::map<std::string, uint32_t> kRootlessConfigFlags = {
+    // CSR_ALLOW_UNTRUSTED_KEXTS
+    {"Untrusted Kernel Extensions", (1 << 0)},
```
Comment (theopolis) [inline osquery/tables/system/darwin/sip_config.cpp]: What about making these more system-parsable, e.g., `allow_untrusted_kexts`, `allow_unrestricted_fs`, `allow_task_for_pid`?

File: osquery/tables/system/darwin/sip_config.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  csr_get_active_config(&config);
+
+  csr_config_t valid_flags = 0;
+  for (const auto& kv : kRootlessConfigFlags) {
+    valid_flags |= kv.second;
+  }
+
+  Row r;
+
+  r["config_flag"] = "System Integrity Protection (SIP)";
```
Comment (theopolis) [inline osquery/tables/system/darwin/sip_config.cpp]: How about `sip_enabled`?

**Comment by theopolis** [comment]
Looks great, I agree with the separate `sip_exceptions` that reports the filesystem configuration-parsed exception lists. I know there's at least 2, I wonder if it'll be possible to amalgamate them into a single list?

**Comment by theopolis** [comment]
@sharvilshah, would you mind making a reference issue with the details/suggestion to weak import on Darwin? If you don't mind, then include the `// Ref: ####` as a comment inline with the code. We can use that as a recommendation moving forward for useful-but-private symbol usage.

**Comment by sharvilshah** [comment]
Thanks for the review @theopolis! Will get it to it later tonight.

Also will pick your brain on what we should consider `enable` and `disable`, which may not exactly be what Apple / `csrutil` considers.

PS: After long and hard search, I know why it's called `csr` : it stands for configurable security restrictions.

**Comment by theopolis** [comment]
Mystery solved! and @osqueryer, test this please

**Comment by sharvilshah** [comment]
Jenkins is throwing the undefined symbols for 10.9. I will have to read the Apple docs more carefully.

**Comment by osqueryer** [comment]
Build (ffed82d1e57740ff6ea72a1816d421b03d8a162a) failed :-1:

**Comment by theopolis** [comment]
My advice is to wrap the whole thing in a preprocessor define check. We define a bunch of custom variables to indicate the build. Then this wont build on 10.9, and a 10.11-built package should follow the weak linking guarantees Apple is making.

**Comment by sharvilshah** [comment]
Aha, it seems that with `weak_import` the symbol(s) might be missing at runtime but still must be found at build time unless one forces the linker to either `suppress` or do a `dynamic_lookup` of all undefined symbols.

Your way sounds easier.

**Comment by sharvilshah** [comment]
Updated.

Now it also parses the `csr-active-config` IOKit key in addition to doing the `csr_check`.
Here's how it looks:

```
osquery> .all sip_config;
+----------------------------+---------+----------------+
| config_flag                | enabled | enabled_in_efi |
+----------------------------+---------+----------------+
| sip_enabled                | 0       | 0              |
| allow_apple_internal       | 0       | 1              |
| allow_device_configuration | 0       | 0              |
| allow_kernel_debugger      | 0       | 0              |
| allow_task_for_pid         | 1       | 1              |
| allow_unrestricted_dtrace  | 1       | 1              |
| allow_unrestricted_fs      | 1       | 1              |
| allow_unrestricted_nvram   | 1       | 1              |
| allow_untrusted_kexts      | 1       | 1              |
+----------------------------+---------+----------------+
```

For reference, in XNU the [CSR flags are defined here](https://opensource.apple.com/source/xnu/xnu-3248.20.55/bsd/sys/csr.h) and `csr_get_active_config` and `csr_check` are [implemented here](https://opensource.apple.com/source/xnu/xnu-3248.20.55/bsd/kern/kern_csr.c)

**Comment by theopolis** [comment]
This looks fantastic! From the table spec perspective, how about `enabled_nvram` or `nvram_enabled`? My two thoughts are, (1) NVRAM is more consistent with Apple's diction and flows nicely with our existing  tables, that's not to say we won't eventually have an efi_variables table; (2) I'm partial to smaller column names with as-succinct-as-possible meaning. What are your thoughts?

**Comment by sharvilshah** [comment]
Agree on both points, updated.

**Comment by theopolis** [comment]
@osqueryer, why aren't you testing this, test this please

**Comment by osqueryer** [comment]
Build (ffed82d1e57740ff6ea72a1816d421b03d8a162a) failed :-1:

**Pull request merged**